trie/branches.py

```diff
@@ -0,0 +1,167 @@
+    BinaryTrie,
+)
+from trie.exceptions import (
+    InvalidKeyError,
+)
+from trie.constants import (
+    BLANK_HASH,
+    KV_TYPE,
+    BRANCH_TYPE,
+    LEAF_TYPE,
+    BYTE_0,
+)
+from trie.validation import (
+    validate_is_bytes,
+)
+
+from trie.utils.sha3 import (
+    keccak,
+)
+from trie.utils.binaries import (
+    encode_to_bin,
+)
+from trie.utils.nodes import (
+    parse_node,
+)
+
+
+def if_branch_exist(db, root_hash, key_prefix):
+    """
+    Given a key prefix, return whether this prefix is
+    the prefix of an existing key in the trie.
+    """
+    validate_is_bytes(key_prefix)
+
+    return _if_branch_exist(db, root_hash, encode_to_bin(key_prefix))
+
+
+def _if_branch_exist(db, node_hash, key_prefix):
+    # Empty trie
+    if node_hash == BLANK_HASH:
+        return False
+    nodetype, left_child, right_child = parse_node(db[node_hash])
+    if nodetype == LEAF_TYPE:
+        if key_prefix:
+            return False
+        return True
+    elif nodetype == KV_TYPE:
+        if not key_prefix:
+            return True
+        if len(key_prefix) < len(left_child):
+            if key_prefix == left_child[:len(key_prefix)]:
+                return True
+            return False
+        else:
+            if key_prefix[:len(left_child)] == left_child:
+                return _if_branch_exist(db, right_child, key_prefix[len(left_child):])
+            return False
+    elif nodetype == BRANCH_TYPE:
+        if not key_prefix:
+            return True
+        if key_prefix[:1] == BYTE_0:
+            return _if_branch_exist(db, left_child, key_prefix[1:])
+        else:
+            return _if_branch_exist(db, right_child, key_prefix[1:])
+
+
+def get_branch(db, root_hash, key):
+    """
+    Get a long-format Merkle branch
+    """
+    validate_is_bytes(key)
+
+    return _get_branch(db, root_hash, encode_to_bin(key))
+
+
+def _get_branch(db, node_hash, keypath):
+    if node_hash == BLANK_HASH:
+        return []
+    node = db[node_hash]
+    nodetype, left_child, right_child = parse_node(node)
+    if nodetype == LEAF_TYPE:
+        if not keypath:
+            return [node]
+        else:
+            raise InvalidKeyError("Key too long")
+    elif nodetype == KV_TYPE:
+        if not keypath:
+            raise InvalidKeyError("Key too short")
+        if keypath[:len(left_child)] == left_child:
+            return [node] + _get_branch(
+                db,
+                right_child,
+                keypath[len(left_child):]
+            )
+        else:
+            raise InvalidKeyError("Key does not exist")
+    elif nodetype == BRANCH_TYPE:
+        if not keypath:
+            raise InvalidKeyError("Key too short")
+        if keypath[:1] == BYTE_0:
+            return [node] + _get_branch(db, left_child, keypath[1:])
+        else:
+            return [node] + _get_branch(db, right_child, keypath[1:])
+
+
+def verify_branch(branch, root_hash, key, value):
+    validate_is_bytes(key)
```

Comment:
This validation can be removed since it occurs within `BinaryTrie.get(...)`.

However, I think that it might be worth validating that `branch`, `value` are well formed since they don't get validated anywhere.  (`root_hash` is validated with `BinaryTrie.__init__(...)`

Reply:
There are a few situations I'm not sure if we should check for.
1. node data not in db => a. a dictionary `KeyError` or b. check if key exist first and raise `NodeNotExistError`?
1.b looks like a more clear error message than 1.a to me

2. get branch from an empty trie => a. return empty branch or b. raise trying to fetch from empty trie error?
Since fetching non-exist key from empty trie returns value `None`, 2.a is probably a more consistent approach with this pattern. Or we can implement them both as 2.b. The question I think is, should the user be awarded of that they are fetching values from an empty trie. (In HexaryTrie it will return `BLANK_NODE` but in BinaryTrie there's no `BLANK_NODE`. And no node hash should be `BLANK_HASH` except root hash.) If we implement 2.a then an empty branch would be a valid branch which seems weird to me but has no significant side effect in my knowledge.

Reply:
On a second thought it's probably not a good idea to raise error while fetching value from empty trie. I think I'm just going to treat empty branch as a valid one.

Reply:
I'm not familiar enough with the binary trie spec to know which of these is right.  

It sounds like treating empty branches as valid violates the [Principle of Least Surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) indicating that raising an exception in this case would be the expected behavior.